contracts/V4Router.sol

```diff
@@ -0,0 +1,189 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.19;
+
+import {Hooks} from "@uniswap/v4-core/contracts/libraries/Hooks.sol";
+import {IPoolManager} from "@uniswap/v4-core/contracts/interfaces/IPoolManager.sol";
+import {ILockCallback} from "@uniswap/v4-core/contracts/interfaces/callback/ILockCallback.sol";
+import {BalanceDelta} from "@uniswap/v4-core/contracts/types/BalanceDelta.sol";
+import {PoolKey} from "@uniswap/v4-core/contracts/types/PoolKey.sol";
+import {Currency, CurrencyLibrary} from "@uniswap/v4-core/contracts/types/Currency.sol";
+import {TickMath} from "@uniswap/v4-core/contracts/libraries/TickMath.sol";
+import {IHooks} from "@uniswap/v4-core/contracts/interfaces/IHooks.sol";
+import {PathKey} from "./libraries/PathKey.sol";
+import {IV4Router} from "./interfaces/IV4Router.sol";
+
+/// @title UniswapV4Routing
+/// @notice Abstract contract that contains all internal logic needed for routing through Uniswap V4 pools
+abstract contract V4Router is IV4Router, ILockCallback {
+    using CurrencyLibrary for Currency;
+
+    IPoolManager immutable poolManager;
+
+    /// @dev Only the pool manager may call this function
+    modifier poolManagerOnly() {
+        if (msg.sender != address(poolManager)) revert NotPoolManager();
+        _;
+    }
+
+    constructor(IPoolManager _poolManager) {
+        poolManager = _poolManager;
+    }
+
+    function _v4Swap(SwapType swapType, bytes memory params) internal {
+        poolManager.lock(abi.encode(SwapInfo(swapType, msg.sender, params)));
+    }
```

Comment:
What's the motivation to having this contract be abstract and not having this be the entrypoint?